# News Classification using Machine Learning

## Overview
This project builds a machine learning model to classify news articles into categories based on their textual content. The goal is to explore how natural language processing (NLP) techniques can be used to automatically understand and organize large collections of text data.

## Research Question
Can machine learning models effectively classify news articles into categories based on textual content?

In [7]:
!pip install datasets scikit-learn

import pandas as pd
from datasets import load_dataset
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score

In [8]:
#Load Dataset
dataset = load_dataset("ag_news")

data = dataset["train"].shuffle(seed=42).select(range(2000))

texts = data["text"]
labels = data["label"]

df = pd.DataFrame({"text": texts, "label": labels})
df.head()


,text,label
0,Bangladesh paralysed by strikes Opposition act...,0
1,Desiring Stability Redskins coach Joe Gibbs ex...,1
2,Will Putin #39;s Power Play Make Russia Safer?...,0
3,U2 pitches for Apple New iTunes ads airing dur...,3
4,S African TV in beheading blunder Public broad...,0


## Dataset
The dataset used is the AG News dataset, accessed via the Hugging Face datasets library.

It contains news articles categorized into four classes:
- World
- Sports
- Business
- Sci/Tech


For this project, the task is simplified into a binary classification problem by grouping the original four categories into two broader classes.
The original four classes are grouped into two categories by combining labels 0–1 into one class and labels 2–3 into another.

In [9]:
#Preprocess and Train
# Convert labels into binary classes (category grouping)
df["label"] = df["label"].apply(lambda x: 0 if x < 2 else 1)

X_train, X_test, y_train, y_test = train_test_split(
    df["text"], df["label"], test_size=0.2, random_state=42
)

vectorizer = TfidfVectorizer(max_features=5000)
X_train_vec = vectorizer.fit_transform(X_train)
X_test_vec = vectorizer.transform(X_test)

model = LogisticRegression()
model.fit(X_train_vec, y_train)

preds = model.predict(X_test_vec)

print("Accuracy:", accuracy_score(y_test, preds))

Accuracy: 0.935


In [10]:
#Predictions
sample_df = pd.DataFrame({
    "text": X_test[:5],
    "true_label": y_test[:5],
    "predicted": preds[:5]
})

sample_df

,text,true_label,predicted
1860,Edmonds back in the swing The last two weeks o...,0,0
353,Revelations Only Confirm Suspicions About Drug...,0,1
1333,Delta Pilots Tell Negotiators to Get Pact NEW...,1,1
905,Britain Agrees to U.S. Troops Request (AP) AP ...,0,0
1289,Supreme Court Affirms Do-Not-Call Ruling The n...,1,0


## Results
The model achieves high accuracy (0.935) in classifying news articles into grouped categories.

Some errors occur due to ambiguity and overlap between certain topics, but overall performance is strong. The high accuracy suggests that even simple models like Logistic Regression combined with TF-IDF can perform well on structured text classification tasks.

## Interpretation
The model demonstrates that machine learning can effectively classify news articles based on their content.

High accuracy indicates that different news categories contain distinct linguistic patterns that can be captured using simple NLP techniques.

## Limitations
- Short text samples
- Simplified binary classification
- No deep learning models used

## Future Work
- Use transformer models for improved accuracy
- Incorporate longer text and metadata
- Explore explainability techniques